napari/utils/translations.py

```diff
@@ -0,0 +1,456 @@
+        func = globals().get(func_name, None)
+
+        if func:
+            try:
+                data, message = func()
+            except Exception:
+                message = traceback.format_exc()
+    else:
+        message = "Invalid number of arguments!"
+
+    sys.stdout.write(json.dumps({"data": data, "message": message}))
+
+
+def _get_installed_language_pack_locales() -> dict:
+    """
+    Get available installed language pack locales.
+
+    Returns
+    -------
+    dict
+        A dict with all the languages found.
+
+    Notes
+    -----
+    This functions are meant to be called via a subprocess to guarantee the
+    results represent the most up-to-date entry point information, which
+    is defined on interpreter startup.
+    """
+    data = {}
+    messages = []
+    for entry_point in pkg_resources.iter_entry_points(
+        NAPARI_LANGUAGEPACK_ENTRY
+    ):
+        try:
+            data[entry_point.name] = os.path.basename(
+                os.path.dirname(entry_point.load().__file__)
+            )
+        except Exception:
+            messages.append(traceback.format_exc())
+
+    message = "\n".join(messages)
+    return data, message
+
+
+def _get_display_name(
+    locale: str, display_locale: str = DEFAULT_LOCALE
+) -> str:
+    """
+    Return the language name to use with a `display_locale` for a given language locale.
+    Parameters
+    ----------
+    locale: str
+        The language name to use.
+    display_locale: str, optional
+        The language to display the `locale`.
+    Returns
+    -------
+    str
+        Localized `locale` and capitalized language name using `display_locale` as language.
+    """
+    # This is a dependency of the language packs to keep out of core
+    import babel
+
+    locale = locale if _is_valid_locale(locale) else DEFAULT_LOCALE
+    display_locale = (
+        display_locale if _is_valid_locale(display_locale) else DEFAULT_LOCALE
+    )
+    loc = babel.Locale.parse(locale)
+    return loc.get_display_name(display_locale).capitalize()
+
+
+def _is_valid_locale(locale: str) -> bool:
+    """
+    Check if a `locale` value is valid.
+
+    Parameters
+    ----------
+    locale: str
+        Language locale code.
+
+    Notes
+    -----
+    A valid locale is in the form language (See ISO-639 standard) and an
+    optional territory (See ISO-3166 standard).
+
+    Examples of valid locales:
+    - English: DEFAULT_LOCALE
+    - Australian English: "en_AU"
+    - Portuguese: "pt"
+    - Brazilian Portuguese: "pt_BR"
+
+    Examples of invalid locales:
+    - Australian Spanish: "es_AU"
+    - Brazilian German: "de_BR"
+    """
+    # This is a dependency of the language packs to keep out of core
+    import babel
+
+    valid = False
+    try:
+        babel.Locale.parse(locale)
+        valid = True
+    except babel.core.UnknownLocaleError:
+        pass
+    except ValueError:
+        pass
+
+    return valid
+
+
+def _run_process_and_parse(cmd: list):
```

Comment:
what's the motivation for the subprocess here?  (is it to avoid the slowness of pkg_resources?).  If it _is_ to achieve some level of concurrency, are there any potential race conditions in qt_main_window using `trans._`?

Reply:
>what's the motivation for the subprocess here? (is it to avoid the slowness of pkg_resources?). If it is to achieve some level of concurrency, are there any potential race conditions in 

No, it is to be able to discover a langage-pack if it was installed after Napari is executed. Python entry points are "pre-discovered" when the python interpreter is instarted so:

If someone open napari, then installs a language pack in the ipyconsole, or another terminal and open preferences, it will not display this newly installed package.


>qt_main_window using trans._?

🤷🏼

Reply:
> Python entry points are "pre-discovered" when the python interpreter is started 

we successfully detect plugins entry points that are installed via the gui, without restarting... so I'm not sure that's the case?  

please have a look at https://github.com/goanpeca/napari/pull/1, which solves all of my concerns... and tell me what you think.